sys/net/network_layer/ng_ipv6/ng_ipv6.c

```diff
@@ -0,0 +1,521 @@
+#include "cpu-conf.h"
+#include "kernel_types.h"
+#include "net/ng_netbase.h"
+#include "net/ng_protnum.h"
+#include "thread.h"
+#include "utlist.h"
+
+#include "net/ng_ipv6/nc.h"
+#include "net/ng_ipv6/netif.h"
+
+#include "net/ng_ipv6.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define _MAX_L2_ADDR_LEN    (8U)
+
+static char _stack[NG_IPV6_STACK_SIZE];
+static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+
+#if ENABLE_DEBUG
+static char addr_str[NG_IPV6_ADDR_MAX_STR_LEN];
+#endif
+
+/* handles NG_NETAPI_MSG_TYPE_RCV commands */
+static void _receive(ng_pktsnip_t *pkt);
+/* dispatches received IPv6 packet for upper layer */
+static void _dispatch_rcv_pkt(ng_nettype_t type, uint32_t demux_ctx,
+                              ng_pktsnip_t *pkt);
+/* Sends packet over the appropriate interface(s). */
+static void _send(ng_pktsnip_t *pkt, bool prep_hdr);
+/* Main event loop for IPv6 */
+static void *_event_loop(void *args);
+
+kernel_pid_t ng_ipv6_init(void)
+{
+    if (_pid == KERNEL_PID_UNDEF) {
+        _pid = thread_create(_stack, NG_IPV6_STACK_SIZE, NG_IPV6_PRIO,
+                             CREATE_STACKTEST, _event_loop, NULL, "ipv6");
+    }
+
+    return _pid;
+}
+
+void ng_ipv6_demux(kernel_pid_t iface, ng_pktsnip_t *pkt, uint8_t nh)
+{
+    int receiver_num;
+
+    pkt->type = ng_nettype_from_protnum(nh);
+
+    /* TODO: add ICMPv6 and extension header handling */
+    (void)iface;    /* will be used by that */
+
+    receiver_num = ng_netreg_num(pkt->type, NG_NETREG_DEMUX_CTX_ALL) +
+                   ng_netreg_num(NG_NETTYPE_IPV6, nh);
+
+    if (receiver_num == 0) {
+        DEBUG("ipv6: unable to forward packet as no one is interested in it\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    ng_pktbuf_hold(pkt, receiver_num - 1);
+    /* IPv6 is not interested anymore so `- 1` */
+    _dispatch_rcv_pkt(pkt->type, NG_NETREG_DEMUX_CTX_ALL, pkt);
+    _dispatch_rcv_pkt(NG_NETTYPE_IPV6, nh, pkt);
+}
+
+/* internal functions */
+static void *_event_loop(void *args)
+{
+    msg_t msg, reply, msg_q[NG_IPV6_MSG_QUEUE_SIZE];
+    ng_netreg_entry_t me_reg;
+
+    (void)args;
+    msg_init_queue(msg_q, NG_IPV6_MSG_QUEUE_SIZE);
+
+    me_reg.demux_ctx = NG_NETREG_DEMUX_CTX_ALL;
+    me_reg.pid = thread_getpid();
+
+    /* register interest in all IPv6 packets */
+    ng_netreg_register(NG_NETTYPE_IPV6, &me_reg);
+
+    /* preinitialize ACK */
+    reply.type = NG_NETAPI_MSG_TYPE_ACK;
+
+    /* start event loop */
+    while (1) {
+        DEBUG("ipv6: waiting for incoming message.\n");
+        msg_receive(&msg);
+
+        switch (msg.type) {
+            case NG_NETAPI_MSG_TYPE_RCV:
+                DEBUG("ipv6: NG_NETAPI_MSG_TYPE_RCV received\n");
+                _receive((ng_pktsnip_t *)msg.content.ptr);
+                break;
+
+            case NG_NETAPI_MSG_TYPE_SND:
+                DEBUG("ipv6: NG_NETAPI_MSG_TYPE_SND received\n");
+                _send((ng_pktsnip_t *)msg.content.ptr, true);
```

Comment:
(see https://github.com/RIOT-OS/RIOT/pull/2454/files#diff-c0ab51e027bbe0cfdb3482d3e29d6f34R320) for documentation. Will move it to the top.